### PR TITLE
[SPARK-33935][SQL] Fix CBO cost function

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Changed the cost function in CBO to match documentation.

### Why are the changes needed?

The parameter `spark.sql.cbo.joinReorder.card.weight` is documented as:
```
The weight of cardinality (number of rows) for plan cost comparison in join reorder: rows * weight + size * (1 - weight).
```
The implementation in `JoinReorderDP.betterThan` does not match this documentaiton:
```    
def betterThan(other: JoinPlan, conf: SQLConf): Boolean = {
      if (other.planCost.card == 0 || other.planCost.size == 0) {
        false
      } else {
        val relativeRows = BigDecimal(this.planCost.card) / BigDecimal(other.planCost.card)
        val relativeSize = BigDecimal(this.planCost.size) / BigDecimal(other.planCost.size)
        relativeRows * conf.joinReorderCardWeight +
          relativeSize * (1 - conf.joinReorderCardWeight) < 1
      }
    }
```

This different implementation has an unfortunate consequence:
given two plans A and B, both A betterThan B and B betterThan A might give the same results. This happes when one has many rows with small sizes and other has few rows with large sizes.

A example values, that have this fenomen with the default weight value (0.7):
A.card = 500, B.card = 300
A.size = 30, B.size = 80
Both A betterThan B and B betterThan A would have score above 1 and would return false.

This happens with several of the TPCDS queries.

The new implementation does not have this behavior.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New and existing UTs

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 1]
```diff
@@ -349,14 +349,11 @@ object JoinReorderDP extends PredicateHelper with Logging {
     }
 
     def betterThan(other: JoinPlan, conf: SQLConf): Boolean = {
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala:351]: This is the only change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 2]
```diff
@@ -349,14 +349,11 @@ object JoinReorderDP extends PredicateHelper with Logging {
...
-        val relativeSize = BigDecimal(this.planCost.size) / BigDecimal(other.planCost.size)
-        relativeRows * conf.joinReorderCardWeight +
-          relativeSize * (1 - conf.joinReorderCardWeight) < 1
-      }
+      val oldCost = BigDecimal(this.planCost.card) * conf.joinReorderCardWeight +
+        BigDecimal(this.planCost.size) * (1 - conf.joinReorderCardWeight)
+      val newCost = BigDecimal(other.planCost.card) * conf.joinReorderCardWeight +
+        BigDecimal(other.planCost.size) * (1 - conf.joinReorderCardWeight)
+      newCost < oldCost
     }
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala:357]: This is the only change

**Comment by tanelk** [comment]
A bit more information:

The unstability of CBO has been noted before (#29638) and I think, that this is the main reason for this. There is also #29871, that tackles another reason for the unstability. That one should only impact, when the costs are equal, this here can impact more plans (see the example in the description)

I'm aware that the UTs will fail because the plans will be different. Because it would be quite a lot of work, I will fix those iff I get some feedback, that this issue is worth looking at further. 

An important thing to note is that this could change the the behavior of the `spark.sql.cbo.joinReorder.card.weight` config value, but luckily it seems, that it does so minimally. 
I generated random values for the plan row counts and sizes, and found that the new cost function agrees most with the old cost function at the same weight value. This holds true for all the weight value, not only the default (0.7).
2020-12-29-195948_1920x1080_scrot

**Comment by tanelk** [comment]
@LuciferYang and @maropu 
Could you take a look at this or perhaps point this to someone, who has worked more with the CBO.

**Comment by maropu** [comment]
cc: @wzhfy

**Comment by maropu** [comment]
I've checked the update and it seems fine. Could you add some tests based on the example in the PR description?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 3]
```diff
@@ -349,14 +349,11 @@ object JoinReorderDP extends PredicateHelper with Logging {
...
-        false
-      } else {
-        val relativeRows = BigDecimal(this.planCost.card) / BigDecimal(other.planCost.card)
-        val relativeSize = BigDecimal(this.planCost.size) / BigDecimal(other.planCost.size)
-        relativeRows * conf.joinReorderCardWeight +
-          relativeSize * (1 - conf.joinReorderCardWeight) < 1
-      }
+      val oldCost = BigDecimal(this.planCost.card) * conf.joinReorderCardWeight +
+        BigDecimal(this.planCost.size) * (1 - conf.joinReorderCardWeight)
+      val newCost = BigDecimal(other.planCost.card) * conf.joinReorderCardWeight +
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala:354]: nit: `newCost` -> `otherCost`?

**Comment by tanelk** [comment]
> I've checked the update and it seems fine. Could you add some tests based on the example in the PR description?

Ahh, good point. Adding this I also realised, that I had the inequality check in the wrong order - this greatly reduced the amount of plan changes.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/joinReorder/JoinReorderSuite.scala [hunk 4]
```diff
@@ -363,4 +364,18 @@ class JoinReorderSuite extends JoinReorderPlanTestBase with StatsEstimationTestB
...
+    val plan2 = JoinPlan(null, null, null, Cost(500, 30))
+
+    // cost1 = 300*0.7 + 80*0.3 = 234
+    // cost2 = 500*0.7 + 30*0.3 = 359
+
+    assert(!plan1.betterThan(plan1, conf))
+    assert(!plan2.betterThan(plan2, conf))
+
+    assert(plan1.betterThan(plan2, conf))
+    assert(!plan2.betterThan(plan1, conf))
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/joinReorder/JoinReorderSuite.scala:379]: Thanks for adding this. This fix looks fine to me. cc: @cloud-fan @HyukjinKwon

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by LuciferYang** [comment]
@tanelk Can you help to run UTs in  `sql/catalyst` and `sql/core` in Scala 2.13 with this pr manually ? thx ~

**Comment by maropu** [comment]
Thanks! Merged to master/3.1.

**Comment by maropu** [comment]
> @tanelk Can you help to run UTs in sql/catalyst and sql/core in Scala 2.13 with this pr manually ? thx ~

Could you check this? Have this fix resolved the previous issue, too?

**Comment by maropu** [comment]
@tanelk Could you open a PR to fix it for branch-3.0?

**Comment by tanelk** [comment]
I created the pulls for 3.0 and 2.4: #31042 and #31043

**Comment by tanelk** [comment]
> @tanelk Can you help to run UTs in `sql/catalyst` and `sql/core` in Scala 2.13 with this pr manually ? thx ~

The `sql/catalyst` passed.
With `sql/core` I had some issues:
at random tests it would fail with the following exception:
```
*** RUN ABORTED ***
  java.net.BindException: Cannot assign requested address: Service 'sparkDriver' failed after 100 retries (on a random free port)!
```

I assume you want the results of `*PlanStability*Suite` - these all passed.

**Comment by maropu** [comment]
> The sql/catalyst passed.
> I assume you want the results of *PlanStability*Suite - these all passed.

Awesome!

**Comment by wzhfy** [comment]
@tanelk Hi, sorry to see this so late.  

The reason to use a relative value for rowCount and size, is to normalize them in a similar scale while comparing cost. Otherwise, one (size) may overwhelm the other (rowCount), then the weight and cost function become meaningless.  

To resolve stability problem, can we have a `betterPlan()` function instead of current `betterThan()`? Inside that plan, we can fix the comparison order based on rowCount or size. And in caller side, we can use `!existingPlan.get.eq(betterPlan(existingPlan.get, newJoinPlan, conf))` to decide whether to update the best plan so far.
```
    def betterPlan(existing: JoinPlan, newPlan: JoinPlan, conf: SQLConf): JoinPlan = {
      // To fix the comparison order, set the one with smaller cardinality as the baseline.
      val (baseline, toCompare) = if (existing.planCost.card <= newPlan.planCost.card) {
        (existing, newPlan)
      } else {
        (newPlan, existing)
      }

      if (toCompare.planCost.card == 0 || toCompare.planCost.size == 0) {
        return existing
      }

      val relativeRows = BigDecimal(baseline.planCost.card) / BigDecimal(toCompare.planCost.card)
      val relativeSize = BigDecimal(baseline.planCost.size) / BigDecimal(toCompare.planCost.size)
      val relativeCost = relativeRows * conf.joinReorderCardWeight +
        relativeSize * (1 - conf.joinReorderCardWeight)
      if (relativeCost == 1) {
        // If they have same cost, we don't update the best plan and return the existing one.
        existing
      } else if (relativeCost < 1) {
        baseline
      } else {
        toCompare
      }
    }
```
What do you think?

**Comment by cloud-fan** [comment]
I think what @wzhfy said makes sense, @tanelk do you have time to try this idea?

**Comment by tanelk** [comment]
@wzhfy and @cloud-fan 

I'm not a fan of adding up the relative costs.

A simple example, where the weight is 0.5:
If this plans size (bytes) is 2x larger, then no matter how many times more rows does the other plan have, the other plan will allways be considered to be better - `0.5*2 + 0.5*0.00000000000001  > 1`.
This basically the same situation, where one cost overwhelms the other.

Perhaps this would be a best of both worlds:
`(this.card / other.card) ^ cardWeight * (this.size / other.size) ^ (1 - cardWeight) < 1`.
In short - multiply the relative costs instead of adding them.

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt [hunk 5]
```diff
@@ -6,71 +6,71 @@ TakeOrderedAndProject [ext_price,brand,brand_id,i_manufact_id,i_manufact]
...
-                                BroadcastHashJoin [ss_store_sk,s_store_sk]
-                                  Project [ss_customer_sk,ss_store_sk,ss_ext_sales_price,i_brand_id,i_brand,i_manufact_id,i_manufact]
-                                    BroadcastHashJoin [ss_sold_date_sk,d_date_sk]
-                                      Project [ss_sold_date_sk,ss_customer_sk,ss_store_sk,ss_ext_sales_price,i_brand_id,i_brand,i_manufact_id,i_manufact]
-                                        BroadcastHashJoin [ss_item_sk,i_item_sk]
-                                          Filter [ss_sold_date_sk,ss_item_sk,ss_customer_sk,ss_store_sk]
-                                            ColumnarToRow
-                                              InputAdapter
-                                                Scan parquet default.store_sales [ss_sold_date_sk,ss_item_sk,ss_customer_sk,ss_store_sk,ss_ext_sales_price]
+                BroadcastHashJoin [ss_item_sk,i_item_sk]
```
Comment (cloud-fan) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: I think q19 exposes a problem. Previously this `BroadcastHashJoin` is run before the `SortMergeJoin`, which reduces the input data of shuffle, because this  `BroadcastHashJoin` has a filter on the right side and can likely prune many data.

@tanelk , if the idea from @wzhfy doesn't look good to you, can you try with some other ideas and see if we can fix this issue?

Comment (tanelk) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: I'll experiment with it a bit, but it might take a while.

Comment (cloud-fan) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: Thanks for looking into it!

Comment (HyukjinKwon) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: This actually caused a significant regression at q19 in TPC-DS benchmark (performed internally). Can we just revert it for now, and do it with actual performance numbers? I think that's safer and easier for everybody here. Seems like At least the plan change here is overlooked and the performance has to be clarified.

Comment (maropu) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: branch-2.4 has the same regression? This PR was back-ported into branch-2.4, too. cc: @viirya

Comment (viirya) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: +1 for @HyukjinKwon's suggestion.

Comment (HyukjinKwon) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: I discussed with @cloud-fan offline. Shall we merge #32014 into master, branch-3.1 and branch-3.0 to fix the regression, and revert this one from branch-2.4? Spark 2.4 release is very soon and might be best to stay safe, and technically this was more like an improvement.

Comment (viirya) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: Agree with it. I prefer to make stable for branch-2.4.

Comment (cloud-fan) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: yea let's revert it from 2.4 to be safe

Comment (viirya) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q19.sf100/simplified.txt:9]: https://github.com/apache/spark/pull/32020
